StatusView component for exposing SDK status messages

tomaszrybakiewicz opened the pull request:
### Description
Closes #5380

This PR introduces `StatusView` that can be used to render Nav SDK status messages.
View allows customization of:

- background Drawable
- message TextAppearance
- spinner Drawable and Tint
- icon Drawable and Tint
- show and hide animations

### Changelog

```
<changelog>Added new `StatusView` component for displaying SDK status messages.</changelog>
```

### Screenshots or Gifs

https://user-images.githubusercontent.com/2678039/150590019-f2003e98-e1de-47c8-adaa-7d4cf4d0e7cf.mp4

abhishek1508 commented on libnavui-status/src/main/res/layout/mapbox_status_view_layout.xml:
```diff
@@ -0,0 +1,64 @@
+            app:layout_constraintBottom_toBottomOf="@+id/messageText"
+            app:layout_constraintDimensionRatio="1:1"
+            android:layout_marginTop="2dp"
+            android:layout_marginBottom="2dp"
```
![Screen Shot 2022-01-24 at 9 12 51 AM](https://user-images.githubusercontent.com/9770186/150831208-4a1837c9-0dbb-476e-abb9-f16a61c48a42.png)

Just wondering why do we need the `marginTop` and `marginBottom`? With these margin values the spinner is not aligned to the text view on the top.

tomaszrybakiewicz replied:
Just a personal preference. It looked a bit oversized to me so I decreased its size by padding it vertically.
It is aligned to a TextView center line.
Would you like me to change this?

abhishek1508 replied:
I think this is a developer preference. If it's related to the sizing, we can still align the view to the center of text view. At the moment, looks like only bottom is aligned.

abhishek1508 commented on libnavui-status/src/main/res/values/colors.xml:
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+    <color name="mapbox_status_view_background">#4D000000</color>
+    <color name="mapbox_status_view_foreground">@android:color/white</color>
```
```suggestion
    <color name="mapbox_status_view_foreground">@color/colorOnSurface</color>
```
This will enable switching b/w day/night mode seamlessly. Also, we would like to keep using the colors defined in `libnavui-resources` to be consistent unless a really different color is required.

tomaszrybakiewicz replied:
👍  Which colors from `libnavui-resources` should I use for background and foreground?

abhishek1508 commented on libnavui-status/src/main/res/values/colors.xml:
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+    <color name="mapbox_status_view_background">#4D000000</color>
```
```suggestion
    <color name="mapbox_status_view_background">@color/colorSurface</color>
```
This will enable switching b/w day/night mode seamlessly. Also, we would like to keep using the colors defined in `libnavui-resources` to be consistent unless a really different color is required.

tomaszrybakiewicz commented on libnavui-status/src/main/java/com/mapbox/navigation/ui/status/view/MapboxStatusView.kt:
```diff
@@ -0,0 +1,233 @@
+ * View for rendering [Status] information.
+ */
+@Suppress("MemberVisibilityCanBePrivate")
+class MapboxStatusView : FrameLayout {
```
True, it does look redundant but making this view a FrameLayout instead of ConstraintLayout hides implementation details  and prevents developers from misusing it as parent for their views. (FrameLayout is designed to display only a single item).

tomaszrybakiewicz commented on libnavui-status/src/main/java/com/mapbox/navigation/ui/status/view/MapboxStatusView.kt:
```diff
@@ -0,0 +1,233 @@
+    /**
+     * Returns `true` if this view is visible.
+     */
+    val isRendered: Boolean get() = isVisible
```
This is to match our "render" abstraction and hide rendered state implementation.

abhishek1508 commented on libnavui-status/src/main/java/com/mapbox/navigation/ui/status/view/MapboxStatusView.kt:
```diff
@@ -0,0 +1,233 @@
+    /**
+     * Container view that hosts [spinnerProgressBar], [iconImage] and [messageTextView],.
+     */
+    val containerView: ConstraintLayout = binding.container
```
Yeah I thought about it later and realized keeping them public would allow for runtime style changes. This should be good

tomaszrybakiewicz commented:
@abhishek1508 I have addressed PR comments and rebased, and squashed on top of latest `main`.
Please let me know if there is anything else that needs to change.

abhishek1508 commented on libnavui-status/src/test/java/com/mapbox/navigation/ui/status/view/MapboxStatusViewTest.kt:
```diff
@@ -0,0 +1,83 @@
+
+@Config(sdk = [Build.VERSION_CODES.O_MR1])
+@RunWith(RobolectricTestRunner::class)
+class MapboxStatusViewTest {
```
Can we add a test to check the behavior of `cancel` and thus `hide`

@tomaszrybakiewicz

tomaszrybakiewicz replied:
I don't think we will be able to do that. For some reason the onAnimationEnd and onAnimationCancel on hideAnimator listener is not being called when testing with Robolectric.

I tried using coroutines to wait for `isRendered` value update before verifying it but with no luck. Second coroutine times out even with 5sec timeout:

```kotlin
    @Test
    fun `cancel - should hide the view`() = runBlocking {
        val status = Status("message", animated = false, icon = android.R.drawable.ic_secure)
        sut.render(status)
        awaitUntil { sut.isRendered }

        sut.cancel()
        awaitUntil(5000) { !sut.isRendered }
        assertFalse("isVisible", sut.isRendered)

    }

    private suspend fun awaitUntil(timeout: Long = 1000, predicate: () -> Boolean) = withTimeout(timeout) {
        while(true) {
            if (predicate()) return@withTimeout
            delay(50)
        }
    }
```

Any ideas how to test hide anim?

abhishek1508 replied:
```
@Test
    fun `cancel - should hide the view`() {
        val status = Status("message", animated = true, icon = android.R.drawable.ic_secure)
        sut.render(status)
        sut.cancel()

        ShadowLooper.idleMainLooper()

        assertFalse("isVisible", sut.isRendered)
    }
```

tomaszrybakiewicz replied:
Nice. Thanks 🙏.

abhishek1508 commented:
Also, I think we still need to apply the elevation as discussed.

tomaszrybakiewicz commented:
@abhishek1508 Done.

- added missing `cancel` test 
- updated MapboxStatusView to support 4 arg constructor that allows default style injection 
- added `android:elevation` to MapboxStyleStatusView style
- removed redundant `statusViewBackground` stylable attribute in favor of `android:background`

Let me know if anything else is missing.

abhishek1508 reviewed (approved):
💯 
BTW CI is failing

@tomaszrybakiewicz

tomaszrybakiewicz commented:
@abhishek1508 Yeah. Forgot to run metalava after adding new constructor. Should be good now.
I still don't know why instrumentation-tests is failing.

kmadsen commented on libnavui-status/src/main/java/com/mapbox/navigation/ui/status/model/Status.kt:
```diff
@@ -0,0 +1,41 @@
+/**
+ * This class stores information to be displayed by the [MapboxStatusView].
+ */
+data class Status(
```
@tomaszrybakiewicz @abhishek1508 having a public constructor like this will be a major semver if we ever want to add a field to it. For this reason we've been using Builder, Mapper, Processor classes

Zayankovsky replied:
Event with default values adding a new field will be a breaking change, because of the generated `copy` method, which doesn't have default values for Java users.

tomaszrybakiewicz replied:
As long as we provide default values for the new fields we should be fine.
For the Java users we can add backward compatible `copy` method.
eg.
```kotlin
data class Status(
 ...
 val newField: Boolean = false
) {
    fun copy(
        message: String = this.message,
        duration: Long = this.duration,
        animated: Boolean = this.animated,
        spinner: Boolean = this.spinner,
        icon: Int = this.icon
    ) = Status(message, duration, animated, spinner, icon, this.newField)
}
```

Zayankovsky replied:
You also need to add `@JvmOverloads` annotation to the constructor. 
I would still prefer to see this class reimplemented using Builder pattern for consistency with the rest of SDK. We have `BuilderTest` that will check your implementation. Adding a new copy method with each new field doesn't look appealing to me.

abhishek1508 replied:
I think on the UI side of the SDK, we have been using `Builder` only when a UI Component has multiple configuration options. There are 2 different kinds of model classes we have been following on the UI side:
- models which cannot be created by an end user: For eg: `JunctionValue`
- models which can be created by an end user: For eg: `Maneuver`
Both these kind of model classes are simple class with an internal keyword: `class Maneuver internal constructor(...)`, however for the models of 2nd type which an end user can create, we associate a Factory with it: `ManeuverFactory` with an `Experimental` annotation. The purpose of the `Experimental` annotation is to inform the user that there might be in future instances where we can add/remove new properties. Even though, I am not a big fan of it and there is a potential to improve, I would suggest to use the same pattern here to keep the consistency with other UI components.

RingerJK commented on libnavui-status/src/main/java/com/mapbox/navigation/ui/status/view/MapboxStatusView.kt:
```diff
@@ -0,0 +1,238 @@
+    @AnimatorRes
+    var hideAnimRes: Int = 0
+
+    constructor(context: Context) : this(context, null)
```
must ba called `super`, not `this`, here and below

tomaszrybakiewicz replied:
[All android views](https://android.googlesource.com/platform/frameworks/base/+/master/core/java/android/widget) are using `this()` to call lower level constructor.

RingerJK replied:
Yeah, for instance we can take [EditText](https://android.googlesource.com/platform/frameworks/base/+/master/core/java/android/widget/EditText.java). If we don't use `this`, then lose "default" style `com.android.internal.R.attr.editTextStyle` for constructor: 

```java
public EditText(Context context, AttributeSet attrs) {
 this(context, attrs, com.android.internal.R.attr.editTextStyle);
}
```
To prevent this, we should always call `this()`(same reason we also don't use `@JvmOverload`)
